go/pserver/optimizer.go

```diff
@@ -1,42 +1,71 @@
 package pserver
 
 /*
-#include "optimizer.h"
+// TODO(zhihong): move compile flags to cmake go_library
+#cgo pkg-config: protobuf
+#cgo CFLAGS: -I ../../
+#cgo LDFLAGS: /Users/dzh/.go/src/github.com/PaddlePaddle/Paddle/build/go/pserver/cclient/libpaddle_go_optimizer.a -lstdc++
+#include "paddle/optimizer/optimizer.h"
 */
 import "C"
 import (
 	"fmt"
 	"unsafe"
 )
 
-type optimizerType int
-
-const (
-	sgd optimizerType = iota
-)
-
 var nullPtr = unsafe.Pointer(uintptr(0))
 
 type optimizer struct {
 	opt *C.struct_paddle_optimizer
+	// used in GetParam, reconstruct Parameter from optimizer
+	ElementType ElementType
 }
 
-func newOptimizer(t optimizerType, learning_rate float64) *optimizer {
+func cArrayToSlice(p unsafe.Pointer, len int) []byte {
+	if p == nullPtr {
+		return nil
+	}
+
+	// create a Go clice backed by a C array, reference:
+	// https://github.com/golang/go/wiki/cgo#turning-c-arrays-into-go-slices
+	//
+	// Go garbage collector will not interact with this data, need
+	// to be freed properly.
+	return (*[1 << 30]byte)(p)[:len:len]
+}
+
+func newOptimizer(paramWithConfigs ParameterWithConfig) *optimizer {
 	o := &optimizer{}
-	o.opt = C.paddle_create_SGD_optimizer(C.double(learning_rate))
+	p := paramWithConfigs.Param
+	c := paramWithConfigs.Config
+	var cbuffer unsafe.Pointer
+	cbuffer = unsafe.Pointer(&p.Content[0])
+	o.opt = C.paddle_create_optimizer((*C.uchar)(&c[0]), C.int(len(c)),
+		C.paddle_element_type(p.ElementType), cbuffer, C.int(len(p.Content)),
+		(*C.char)(nullPtr), 0)
 	return o
 }
 
-func (o *optimizer) UpdateParameter(p Parameter, g Gradient) error {
-	if len(p.Content) != len(g.Content) {
-		return fmt.Errorf("Name: %s, parameter and gradient length not match, parameter: %d, gradient: %d", p.Name, len(p.Content), len(g.Content))
+func (o *optimizer) GetWeights(p *Parameter) error {
+	// FIXME: get weigths from optimizer has bug
+	var buffer unsafe.Pointer
+	buffer_len := C.paddle_optimizer_get_weights(o.opt, &buffer)
+	if buffer_len == 0 || buffer == nullPtr {
+		return fmt.Errorf("parameter optimizer error : %s get failed", p.Name)
 	}
+	p.Content = cArrayToSlice(buffer, int(buffer_len))
+	return nil
+}
 
-	if p.ElementType != g.ElementType {
-		return fmt.Errorf("Name: %s, parameter and gradient element type not match, parameter: %v, gradient: %v", p.Name, p.ElementType, g.ElementType)
+func (o *optimizer) UpdateParameter(g Gradient) error {
+	if o.ElementType != g.ElementType {
+		return fmt.Errorf("Name: %s, parameter and gradient element type not match, parameter: %v, gradient: %v", g.Name, o.ElementType, g.ElementType)
 	}
 
-	r := C.paddle_update_parameter(o.opt, unsafe.Pointer(&p.Content[0]), C.paddle_element_type(p.ElementType), unsafe.Pointer(&g.Content[0]), C.int(len(g.Content)))
+	// FIXME: do we need a copy? discard g.Content by GC ok
```

Comment:
We do not need to make a copy of `&g.Content[0]`, since `C.paddle_update_parameter` will not use `&g.Content[0]` anymore after return.